### DISCUSSION TITLE
Which module depends on canvas?

### DISCUSSION THREAD

**Discussion by scccboom** [verdaccio/verdaccio#2183] [Q&A] [answered]
```
npm ERR! path D:\Apps\nodejs\node_modules\verdaccio\node_modules\canvas
```
Is it necessary?
It causes the installation of verdaccio to fail, and the experience is very bad！

**Comment by DanielRuf** [comment]
Try `npm explain canvas` or `yarn why canvas`. What does this show?

Also it mught fail because of missing deoendencies like `python2` to compile native binaries. Please check the whole error message / log for details.

**Comment by scccboom** [comment]
```
npm install -g verdaccio
```
npm WARN deprecated har-validator@5.1.5: this
npm WARN deprecated node-pre-gyp@0.15.0: Please upgrade to @mapbox/node-pre-gyp@1.x: the non-scoped 'node-pre-gyp' package is deprecated and only the @mapbox scoped package will recieve updates in the future
npm ERR! code 1
npm ERR! path D:\Apps\nodejs\node_modules\verdaccio\node_modules\canvas
npm ERR! command failed
npm ERR! command C:\Windows\system32\cmd.exe /d /s /c node-gyp rebuild
npm ERR! Warning: Missing input files:
npm ERR! C:\GTK\bin\libfreetype-6.dll
npm ERR! C:\GTK\bin\libgthread-2.0-0.dll
npm ERR! C:\GTK\bin\libcairo-2.dll
npm ERR! C:\GTK\bin\libpangowin32-1.0-0.dll
npm ERR! C:\GTK\bin\libpangocairo-1.0-0.dll
npm ERR! C:\GTK\bin\libfontconfig-1.dll
npm ERR! C:\GTK\bin\libintl-8.dll
npm ERR! C:\GTK\bin\libpangoft2-1.0-0.dll
npm ERR! C:\GTK\bin\libexpat-1.dll
npm ERR! C:\GTK\bin\libglib-2.0-0.dll
npm ERR! C:\GTK\bin\libgobject-2.0-0.dll
npm ERR! C:\GTK\bin\libpango-1.0-0.dll
npm ERR! C:\GTK\bin\libpng14-14.dll
npm ERR! C:\GTK\bin\libgmodule-2.0-0.dll
npm ERR! C:\GTK\bin\zlib1.dll
npm ERR! �ڴ˽��������һ������һ����Ŀ����Ҫ���ò������ɣ������ӡ�-m�����ء�
npm ERR!   Backend.cc
npm ERR! D:\Apps\nodejs\node_modules\verdaccio\node_modules\canvas\src\backend\Backend.h(3,10): fatal error C1083: �޷�������ļ�: ��cairo.h��: No such file or directory [D:\Apps\nodejs\node_modules\verdaccio\node_modules\canvas\build\canvas.vcxproj]
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@7.1.2
npm ERR! gyp info using node@12.18.4 | win32 | x64
npm ERR! gyp info find Python using Python version 3.8.0 found at "D:\Apps\Python38\python.exe"
npm ERR! gyp info find VS using VS2019 (16.9.31025.194) found at:
npm ERR! gyp info find VS "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise"
npm ERR! gyp info find VS run with --verbose for detailed information
npm ERR! gyp info spawn D:\Apps\Python38\python.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'D:\\Apps\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'msvs',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'D:\\Apps\\nodejs\\node_modules\\verdaccio\\node_modules\\canvas\\build\\config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'D:\\Apps\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\xx\\AppData\\Local\\node-gyp\\Cache\\12.18.4\\include\\node\\common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=C:\\Users\\xx\\AppData\\Local\\node-gyp\\Cache\\12.18.4',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=D:\\Apps\\nodejs\\node_modules\\npm\\node_modules\\node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=C:\\\\Users\\\\xx\\\\AppData\\\\Local\\\\node-gyp\\\\Cache\\\\12.18.4\\\\<(target_arch)\\\\node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=D:\\Apps\\nodejs\\node_modules\\verdaccio\\node_modules\\canvas',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'D:\\Apps\\nodejs\\node_modules\\verdaccio\\node_modules\\canvas\\build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp info spawn C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\MSBuild.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'build/binding.sln',
npm ERR! gyp info spawn args   '/clp:Verbosity=minimal',
npm ERR! gyp info spawn args   '/nologo',
npm ERR! gyp info spawn args   '/p:Configuration=Release;Platform=x64'
npm ERR! gyp info spawn args ]
npm ERR! gyp ERR! build error
npm ERR! gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\MSBuild.exe` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onExit (D:\Apps\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
npm ERR! gyp ERR! System Windows_NT 10.0.19042
npm ERR! gyp ERR! command "D:\\Apps\\nodejs\\node.exe" "D:\\Apps\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd D:\Apps\nodejs\node_modules\verdaccio\node_modules\canvas
npm ERR! gyp ERR! node -v v12.18.4
npm ERR! gyp ERR! node-gyp -v v7.1.2
npm ERR! gyp ERR! not ok

**Reply by DanielRuf** [reply]
Looks like some issues with GTK and missing dependencies for graphics (cairo).

Did you try `npm i -g windows-build-tools`? Never had this problem on Windows, but I've also not used a custom GTK and Visual Studio setup.

**Reply by DanielRuf** [reply]
Please see https://github.com/Automattic/node-canvas/wiki/Installation:-Windows#2-installing-gtk-2

**Comment by juanpicado** [comment, marked as answer]
`canvas` is a `peerDependencies` from [`jsdom`](https://github.com/verdaccio/verdaccio/blob/5.x/yarn.lock#L9238). You are using npm@7 as far I see, right? and `peerDependencies` are installed by default even if verdaccio does not need it. You have three options
- Install the compiler as @DanielRuf suggested
- Downgrade npm to v6 (or see thread 🧵 below)
- Use pnpm or yarn which does not install `peerDependencies` by default.

**Reply by DanielRuf** [reply]
One more option:

`npm i -g verdaccio --legacy-peer-deps`

See https://docs.npmjs.com/cli/v7/commands/npm-install

**Reply by juanpicado** [reply]
Thanks @DanielRuf  I totally forgot that option is also valid.

**Comment by scccboom** [comment]
Thanks for your help. I downgrade npm to v6 and installed.
